Dart/src/com/jetbrains/lang/dart/ide/refactoring/DartInlineHandler.java

```diff
@@ -0,0 +1,226 @@
+    final InlineRefactoringContext context = findContext(editor);
+    if (context == null) {
+      return;
+    }
+    // create refactoring
+    final ServerRefactoring refactoring;
+    if (ElementKind.LOCAL_VARIABLE.equals(context.kind)) {
+      refactoring = new ServerInlineLocalRefactoring(context.filePath, context.offset, 0);
+    } else {
+      refactoring = new ServerInlineMethodRefactoring(context.filePath, context.offset, 0);
+    }
+    // validate initial status
+    {
+      final RefactoringStatus initialConditions = refactoring.checkInitialConditions();
+      if (showMessageIfError(editor, initialConditions)) {
+        return;
+      }
+    }
+    // configure using dialog
+    if (refactoring instanceof ServerInlineMethodRefactoring) {
+      boolean dialogOK = new InlineMethodDialog(project, element, (ServerInlineMethodRefactoring)refactoring).showAndGet();
+      if (!dialogOK) {
+        return;
+      }
+    }
+    // validate final status
+    {
+      final RefactoringStatus finalConditions = refactoring.checkFinalConditions();
+      if (showMessageIfError(editor, finalConditions)) {
+        return;
+      }
+    }
+    // Apply the change.
+    ApplicationManager.getApplication().runWriteAction(new Runnable() {
+      @Override
+      public void run() {
+        final SourceChange change = refactoring.getChange();
+        assert change != null;
+        AssistUtils.applySourceChange(project, change);
+      }
+    });
+  }
+
+  @Override
+  public boolean isEnabledForLanguage(Language l) {
+    return l == DartLanguage.INSTANCE;
+  }
+
+  @Override
+  public boolean isEnabledOnElement(PsiElement element, @Nullable Editor editor) {
+    return canInlineElementInEditor(element, editor);
+  }
+
+  @Nullable
+  static private InlineRefactoringContext findContext(@Nullable Editor editor) {
+    if (editor == null) {
+      return null;
+    }
+    // prepare project
+    final Project project = editor.getProject();
+    if (project == null) {
+      return null;
+    }
+    // prepare files
+    final Document document = editor.getDocument();
+    final PsiFile psiFile = PsiDocumentManager.getInstance(project).getPsiFile(document);
+    if (psiFile == null) {
+      return null;
+    }
+    final VirtualFile virtualFile = psiFile.getVirtualFile();
+    final String filePath = virtualFile.getPath();
+    // prepare navigation regions
+    final int offset = editor.getCaretModel().getOffset();
+    final List<PluginNavigationRegion> navigationRegions = DartAnalysisServerService.getInstance().getNavigation(virtualFile);
+    // find the navigation region
+    for (PluginNavigationRegion region : navigationRegions) {
+      if (region.getOffset() <= offset && offset <= region.getOffset() + region.getLength()) {
+        final List<PluginNavigationTarget> targets = region.getTargets();
+        final String kind = targets.get(0).getKind();
```

Comment:
It was undefined.
I'm changing the DAS spec to explicitly document that it cannot be empty.